**PROBLEM STATEMENT**

In this dataset, you are provided with 7398 movies and a variety of metadata obtained from The Movie Database (TMDB). Movies are labeled with id. Data points include cast, crew, plot keywords, budget, posters, release dates, languages, production companies, and countries.

You are predicting the worldwide revenue for 4398 movies in the test file.

Note - many movies are remade over the years, therefore it may seem like multiple instance of a movie may appear in the data, however they are different and should be considered separate movies. In addition, some movies may share a title, but be entirely unrelated.

E.g. The Karate Kid (id: 5266) was released in 1986, while a clearly (or maybe just subjectively) inferior remake (id: 1987) was released in 2010. Also, while the Frozen (id: 5295) released by Disney in 2013 may be the household name, don't forget about the less-popular Frozen (id: 139) released three years earlier about skiers who are stranded on a chairlift...

**Importing** **Libraries**

In [ ]:
import pandas as pd  #used for data analysis and manipulation
import numpy as np   #library used for working with arrays
from sklearn.model_selection import train_test_split #splits arrays or matrices into random subsets
from sklearn.linear_model import LinearRegression   #is a library which contains logistic regression
from sklearn.metrics import mean_squared_error
# !pip install wordcloud opendatasets scikit-learn --upgrade --quiet
import matplotlib.pyplot as plt  #used for plotting data
import seaborn as sns  #used for stats
%matplotlib inline
#displays the output in the current window
sns.set_style('darkgrid')
from datetime import datetime
import datetime
from collections import Counter
# import os
import ast  #abstract syntax trees
pd.set_option("display.max_columns", 120)
pd.set_option("display.max_rows", 120)

import warnings
warnings.filterwarnings("ignore")

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


**Reading** **the data**

In [ ]:
train = pd.read_csv("/content/drive/MyDrive/train.csv")
test = pd.read_csv("/content/drive/MyDrive/test.csv")
submission = pd.read_csv("/content/drive/MyDrive/ sample_submission.csv")



## EDA

In [ ]:
train.head()

id                              belongs_to_collection    budget  \
0   1  [{'id': 313576, 'name': 'Hot Tub Time Machine ...  14000000   
1   2  [{'id': 107674, 'name': 'The Princess Diaries ...  40000000   
2   3                                                NaN   3300000   
3   4                                                NaN   1200000   
4   5                                                NaN         0   

                                              genres  \
0                     [{'id': 35, 'name': 'Comedy'}]   
1  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
2                      [{'id': 18, 'name': 'Drama'}]   
3  [{'id': 53, 'name': 'Thriller'}, {'id': 18, 'n...   
4  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   

                            homepage    imdb_id original_language  \
0                                NaN  tt2637294                en   
1                                NaN  tt0368933                en   
2  http://sonyclassics.com/whiplash/  tt2582802                en   
3         http://kahaanithefilm.com/  tt1821480                hi   
4                                NaN  tt1380152                ko   

                             original_title  \
0                    Hot Tub Time Machine 2   
1  The Princess Diaries 2: Royal Engagement   
2                                  Whiplash   
3                                   Kahaani   
4                                      마린보이   

                                            overview  popularity  \
0  When Lou, who has become the "father of the In...    6.575393   
1  Mia Thermopolis is now a college graduate and ...    8.248895   
2  Under the direction of a ruthless instructor, ...   64.299990   
3  Vidya Bagchi (Vidya Balan) arrives in Kolkata ...    3.174936   
4  Marine Boy is the story of a former national s...    1.148070   

                        poster_path  \
0  /tQtWuwvMf0hCc2QR2tkolwl7c3c.jpg   
1  /w9Z7A0GHEhIp7etpj0vyKOeU1Wx.jpg   
2  /lIv1QinFqz4dlp5U4lQ6HaiskOZ.jpg   
3  /aTXRaPrWSinhcmCrcfJK17urp3F.jpg   
4  /m22s7zvkVFDU9ir56PiiqIEWFdT.jpg   

                                production_companies  \
0  [{'name': 'Paramount Pictures', 'id': 4}, {'na...   
1        [{'name': 'Walt Disney Pictures', 'id': 2}]   
2  [{'name': 'Bold Films', 'id': 2266}, {'name': ...   
3                                                NaN   
4                                                NaN   

                                production_countries release_date  runtime  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...      2/20/15     93.0   
1  [{'iso_3166_1': 'US', 'name': 'United States o...       8/6/04    113.0   
2  [{'iso_3166_1': 'US', 'name': 'United States o...     10/10/14    105.0   
3            [{'iso_3166_1': 'IN', 'name': 'India'}]       3/9/12    122.0   
4      [{'iso_3166_1': 'KR', 'name': 'South Korea'}]       2/5/09    118.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
4           [{'iso_639_1': 'ko', 'name': '한국어/조선말'}]  Released   

                                             tagline  \
0  The Laws of Space and Time are About to be Vio...   
1  It can take a lifetime to find true love; she'...   
2    The road to greatness can take you to the edge.   
3                                                NaN   
4                                                NaN   

                                      title  \
0                    Hot Tub Time Machine 2   
1  The Princess Diaries 2: Royal Engagement   
2                                  Whiplash   
3                                   Kahaani   
4                                Marine Boy   

                                            Keywords  \
0  [{'id': 4379, 'name':

In [ ]:
train['belongs_to_collection']

0       [{'id': 313576, 'name': 'Hot Tub Time Machine ...
1       [{'id': 107674, 'name': 'The Princess Diaries ...
2                                                     NaN
3                                                     NaN
4                                                     NaN
                              ...                        
2995                                                  NaN
2996                                                  NaN
2997                                                  NaN
2998                                                  NaN
2999                                                  NaN
Name: belongs_to_collection, Length: 3000, dtype: object

In [ ]:
train.shape

(3000, 23)

In [ ]:
test.shape

(4398, 22)

I think we can drop the poster path

In [ ]:
train = train.drop(['poster_path'],axis = 1)
test = test.drop(['poster_path'],axis = 1)

The discussion section has the data which was missing or wrong in the real data set 

In [ ]:
train.loc[train['id'] == 16,'revenue'] = 192864          # Skinning
train.loc[train['id'] == 90,'budget'] = 30000000         # Sommersby          
train.loc[train['id'] == 118,'budget'] = 60000000        # Wild Hogs
train.loc[train['id'] == 149,'budget'] = 18000000        # Beethoven
train.loc[train['id'] == 313,'revenue'] = 12000000       # The Cookout 
train.loc[train['id'] == 451,'revenue'] = 12000000       # Chasing Liberty
train.loc[train['id'] == 464,'budget'] = 20000000        # Parenthood
train.loc[train['id'] == 470,'budget'] = 13000000        # The Karate Kid, Part II
train.loc[train['id'] == 513,'budget'] = 930000          # From Prada to Nada
train.loc[train['id'] == 797,'budget'] = 8000000         # Welcome to Dongmakgol
train.loc[train['id'] == 819,'budget'] = 90000000        # Alvin and the Chipmunks: The Road Chip
train.loc[train['id'] == 850,'budget'] = 90000000        # Modern Times
train.loc[train['id'] == 1007,'budget'] = 2              # Zyzzyx Road 
train.loc[train['id'] == 1112,'budget'] = 7500000        # An Officer and a Gentleman
train.loc[train['id'] == 1131,'budget'] = 4300000        # Smokey and the Bandit   
train.loc[train['id'] == 1359,'budget'] = 10000000       # Stir Crazy 
train.loc[train['id'] == 1542,'budget'] = 1              # All at Once
train.loc[train['id'] == 1570,'budget'] = 15800000       # Crocodile Dundee II
train.loc[train['id'] == 1571,'budget'] = 4000000        # Lady and the Tramp
train.loc[train['id'] == 1714,'budget'] = 46000000       # The Recruit
train.loc[train['id'] == 1721,'budget'] = 17500000       # Cocoon
train.loc[train['id'] == 1865,'revenue'] = 25000000      # Scooby-Doo 2: Monsters Unleashed
train.loc[train['id'] == 1885,'budget'] = 12             # In the Cut
train.loc[train['id'] == 2091,'budget'] = 10             # Deadfall
train.loc[train['id'] == 2268,'budget'] = 17500000       # Madea Goes to Jail budget
train.loc[train['id'] == 2491,'budget'] = 6              # Never Talk to Strangers
train.loc[train['id'] == 2602,'budget'] = 31000000       # Mr. Holland's Opus
train.loc[train['id'] == 2612,'budget'] = 15000000       # Field of Dreams
train.loc[train['id'] == 2696,'budget'] = 10000000       # Nurse 3-D
train.loc[train['id'] == 2801,'budget'] = 10000000       # Fracture
train.loc[train['id'] == 335,'budget'] = 2 
train.loc[train['id'] == 348,'budget'] = 12
train.loc[train['id'] == 470,'budget'] = 13000000 
train.loc[train['id'] == 513,'budget'] = 1100000
train.loc[train['id'] == 640,'budget'] = 6 
train.loc[train['id'] == 696,'budget'] = 1
train.loc[train['id'] == 797,'budget'] = 8000000 
train.loc[train['id'] == 850,'budget'] = 1500000
train.loc[train['id'] == 1199,'budget'] = 5 
train.loc[train['id'] == 1282,'budget'] = 9               # Death at a Funeral
train.loc[train['id'] == 1347,'budget'] = 1
train.loc[train['id'] == 1755,'budget'] = 2
train.loc[train['id'] == 1801,'budget'] = 5
train.loc[train['id'] == 1918,'budget'] = 592 
train.loc[train['id'] == 2033,'budget'] = 4
train.loc[train['id'] == 2118,'budget'] = 344 
train.loc[train['id'] == 2252,'budget'] = 130
train.loc[train['id'] == 2256,'budget'] = 1 
train.loc[train['id'] == 2696,'budget'] = 10000000


test.loc[test['id'] == 6733,'budget'] = 5000000
test.loc[test['id'] == 3889,'budget'] = 15000000
test.loc[test['id'] == 6683,'budget'] = 50000000
test.loc[test['id'] == 5704,'budget'] = 4300000
test.loc[test['id'] == 6109,'budget'] = 281756
test.loc[test['id'] == 7242,'budget'] = 10000000
test.loc[test['id'] == 7021,'budget'] = 17540562       #  Two Is a Family
test.loc[test['id'] == 5591,'budget'] = 4000000        # The Orphanage
test.loc[test['id'] == 4282,'budget'] = 20000000       # Big Top Pee-wee
test.loc[test['id'] == 3033,'budget'] = 250 
test.loc[test['id'] == 3051,'budget'] = 50
test.loc[test['id'] == 3084,'budget'] = 337
test.loc[test['id'] == 3224,'budget'] = 4  
test.loc[test['id'] == 3594,'budget'] = 25  
test.loc[test['id'] == 3619,'budget'] = 500  
test.loc[test['id'] == 3831,'budget'] = 3  
test.loc[test['id'] == 3935,'budget'] = 500  
test.loc[test['id'] == 4049,'budget'] = 995946 
test.loc[test['id'] == 4424,'budget'] = 3  
test.loc[test['id'] == 4460,'budget'] = 8  
test.loc[test['id'] == 4555,'budget'] = 1200000 
test.loc[test['id'] == 4624,'budget'] = 30 
test.loc[test['id'] == 4645,'budget'] = 500 
test.loc[test['id'] == 4709,'budget'] = 450 
test.loc[test['id'] == 4839,'budget'] = 7
test.loc[test['id'] == 3125,'budget'] = 25 
test.loc[test['id'] == 3142,'budget'] = 1
test.loc[test['id'] == 3201,'budget'] = 450
test.loc[test['id'] == 3222,'budget'] = 6
test.loc[test['id'] == 3545,'budget'] = 38
test.loc[test['id'] == 3670,'budget'] = 18
test.loc[test['id'] == 3792,'budget'] = 19
test.loc[test['id'] == 3881,'budget'] = 7
test.loc[test['id'] == 3969,'budget'] = 400
test.loc[test['id'] == 4196,'budget'] = 6
test.loc[test['id'] == 4221,'budget'] = 11
test.loc[test['id'] == 4222,'budget'] = 500
test.loc[test['id'] == 4285,'budget'] = 11
test.loc[test['id'] == 4319,'budget'] = 1
test.loc[test['id'] == 4639,'budget'] = 10
test.loc[test['id'] == 4719,'budget'] = 45
test.loc[test['id'] == 4822,'budget'] = 22
test.loc[test['id'] == 4829,'budget'] = 20
test.loc[test['id'] == 4969,'budget'] = 20
test.loc[test['id'] == 5021,'budget'] = 40 
test.loc[test['id'] == 5035,'budget'] = 1 
test.loc[test['id'] == 5063,'budget'] = 14 
test.loc[test['id'] == 5119,'budget'] = 2 
test.loc[test['id'] == 5214,'budget'] = 30 
test.loc[test['id'] == 5221,'budget'] = 50 
test.loc[test['id'] == 4903,'budget'] = 15
test.loc[test['id'] == 4983,'budget'] = 3
test.loc[test['id'] == 5102,'budget'] = 28
test.loc[test['id'] == 5217,'budget'] = 75
test.loc[test['id'] == 5224,'budget'] = 3 
test.loc[test['id'] == 5469,'budget'] = 20 
test.loc[test['id'] == 5840,'budget'] = 1 
test.loc[test['id'] == 5960,'budget'] = 30
test.loc[test['id'] == 6506,'budget'] = 11 
test.loc[test['id'] == 6553,'budget'] = 280
test.loc[test['id'] == 6561,'budget'] = 7
test.loc[test['id'] == 6582,'budget'] = 218
test.loc[test['id'] == 6638,'budget'] = 5
test.loc[test['id'] == 6749,'budget'] = 8 
test.loc[test['id'] == 6759,'budget'] = 50 
test.loc[test['id'] == 6856,'budget'] = 10
test.loc[test['id'] == 6858,'budget'] =  100
test.loc[test['id'] == 6876,'budget'] =  250
test.loc[test['id'] == 6972,'budget'] = 1
test.loc[test['id'] == 7079,'budget'] = 8000000
test.loc[test['id'] == 7150,'budget'] = 118
test.loc[test['id'] == 6506,'budget'] = 118
test.loc[test['id'] == 7225,'budget'] = 6
test.loc[test['id'] == 7231,'budget'] = 85
test.loc[test['id'] == 5222,'budget'] = 5
test.loc[test['id'] == 5322,'budget'] = 90
test.loc[test['id'] == 5350,'budget'] = 70
test.loc[test['id'] == 5378,'budget'] = 10
test.loc[test['id'] == 5545,'budget'] = 80
test.loc[test['id'] == 5810,'budget'] = 8
test.loc[test['id'] == 5926,'budget'] = 300
test.loc[test['id'] == 5927,'budget'] = 4
test.loc[test['id'] == 5986,'budget'] = 1
test.loc[test['id'] == 6053,'budget'] = 20
test.loc[test['id'] == 6104,'budget'] = 1
test.loc[test['id'] == 6130,'budget'] = 30
test.loc[test['id'] == 6301,'budget'] = 150
test.loc[test['id'] == 6276,'budget'] = 100
test.loc[test['id'] == 6473,'budget'] = 100
test.loc[test['id'] == 6842,'budget'] = 30
test.loc[test['id'] == 3829,'release_date'] = '05/01/2000'

In [ ]:
test.head(3)

id                              belongs_to_collection  budget  \
0  3001  [{'id': 34055, 'name': 'Pokémon Collection', '...       0   
1  3002                                                NaN   88000   
2  3003                                                NaN       0   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 16, '...   
1  [{'id': 27, 'name': 'Horror'}, {'id': 878, 'na...   
2  [{'id': 35, 'name': 'Comedy'}, {'id': 10749, '...   

                                            homepage    imdb_id  \
0  http://www.pokemon.com/us/movies/movie-pokemon...  tt1226251   
1                                                NaN  tt0051380   
2                                                NaN  tt0118556   

  original_language               original_title  \
0                ja           ディアルガVSパルキアVSダークライ   
1                en  Attack of the 50 Foot Woman   
2                en             Addicted to Love   

                                            overview  popularity  \
0  Ash and friends (this time accompanied by newc...    3.851534   
1  When an abused wife grows to giant size becaus...    3.559789   
2  Good-natured astronomer Sam is devastated when...    8.085194   

                                production_companies  \
0                                                NaN   
1  [{'name': 'Woolner Brothers Pictures Inc.', 'i...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries release_date  runtime  \
0  [{'iso_3166_1': 'JP', 'name': 'Japan'}, {'iso_...      7/14/07     90.0   
1  [{'iso_3166_1': 'US', 'name': 'United States o...      5/19/58     65.0   
2  [{'iso_3166_1': 'US', 'name': 'United States o...      5/23/97    100.0   

                                    spoken_languages    status  \
0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
1           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline  \
0  Somewhere Between Time & Space... A Legend Is ...   
1  A titanic beauty spreads a macabre wave of hor...   
2          A Comedy About Lost Loves And Last Laughs   

                          title  \
0  Pokémon: The Rise of Darkrai   
1   Attack of the 50 Foot Woman   
2              Addicted to Love   

                                            Keywords  \
0  [{'id': 11451, 'name': 'pokémon'}, {'id': 115...   
1  [{'id': 9748, 'name': 'revenge'}, {'id': 9951,...   
2  [{'id': 931, 'name': 'jealousy'}, {'id': 9673,...   

                                                cast  \
0  [{'cast_id': 3, 'character': 'Tonio', 'credit_...   
1  [{'cast_id': 2, 'character': 'Nancy Fowler Arc...   
2  [{'cast_id': 11, 'character': 'Maggie', 'credi...   

                                                crew  
0  [{'credit_id': '52fe44e7c3a368484e03d683', 'de...  
1  [{'credit_id': '55807805c3a3685b1300060b', 'de...  
2  [{'credit_id': '52fe4330c3a36847f8041367', 'de...

Counting  missing values

In [ ]:
train.isna().sum()

id                          0
belongs_to_collection    2396
budget                      0
genres                      7
homepage                 2054
imdb_id                     0
original_language           0
original_title              0
overview                    8
popularity                  0
production_companies      156
production_countries       55
release_date                0
runtime                     2
spoken_languages           20
status                      0
tagline                   597
title                       0
Keywords                  276
cast                       13
crew                       16
revenue                     0
dtype: int64

In [ ]:
test.isna().sum()

id                          0
belongs_to_collection    3521
budget                      0
genres                     16
homepage                 2978
imdb_id                     0
original_language           0
original_title              0
overview                   14
popularity                  0
production_companies      258
production_countries      102
release_date                0
runtime                     4
spoken_languages           42
status                      2
tagline                   863
title                       3
Keywords                  393
cast                       13
crew                       22
dtype: int64

converting strings to dictionary

In [ ]:
dict_columns = ['belongs_to_collection', 'genres', 'production_companies',
                'production_countries', 'spoken_languages', 'Keywords', 'cast', 'crew']

def text_to_dict(df):
    for column in dict_columns:
        df[column] = df[column].apply(lambda x: {} if pd.isna(x) else ast.literal_eval(x) )
    return df
        
train = text_to_dict(train)
test = text_to_dict(test)

Imputing the missing values for 'runtime' by the average values in the dataset.

In [ ]:
train['runtime']=train['runtime'].fillna(train['runtime'].mean())
test['runtime']=test['runtime'].fillna(test['runtime'].mean())

In [ ]:
sns.jointplot(x='budget',y='revenue',data = train,kind = 'reg', height=11, ratio=4,joint_kws = {'ci' : None});

In [ ]:
sns.jointplot(x='popularity',y='revenue',data = train, height=11, ratio=4, kind = 'reg', joint_kws = {'ci' : None});

In [ ]:
sns.jointplot(x="runtime", y="revenue", data=train, height=11, ratio=4, kind = 'reg', joint_kws = {'ci' : None});

People like to watch movies which have moderate runtime neither too high nor too low

In [ ]:
sns.distplot(train.revenue)

In [ ]:
test['budget']=np.log1p(test['budget'])
train['budget']=np.log1p(train['budget'])

train['revenue']=np.log1p(train['revenue'])

In [ ]:
sns.distplot(train.revenue)

In [ ]:
# splitting the 'release_date' string to month, date and year
train[['release_month','release_day','release_year']]=train['release_date'].str.split('/',expand=True).replace(np.nan, -1).astype(int)
# the dataset consists of data from 1900s to 2017 
# so since we are only goven the year in 2 digits we add the respective centuries (eg : 12 -> 2012 )

train.loc[ (train['release_year'] <= 20) & (train['release_year'] < 100), "release_year"] += 2000
# eg: 75 -> 1975
train.loc[ (train['release_year'] > 20)  & (train['release_year'] < 100), "release_year"] += 1900
# getting the day of week and quarter of the year
releaseDate = pd.to_datetime(train['release_date']) 
train['release_dayofweek'] = releaseDate.dt.dayofweek
train['release_quarter'] = releaseDate.dt.quarter

In [ ]:
# same for the test set
test[['release_month','release_day','release_year']]=test['release_date'].str.split('/',expand=True).replace(np.nan, -1).astype(int)
test.loc[ (test['release_year'] <= 20) & (test['release_year'] < 100), "release_year"] += 2000
test.loc[ (test['release_year'] > 20)  & (test['release_year'] < 100), "release_year"] += 1900
releaseDate = pd.to_datetime(test['release_date']) 
test['release_dayofweek'] = releaseDate.dt.dayofweek
test['release_quarter'] = releaseDate.dt.quarter

In [ ]:
plt.figure(figsize = (18,12))
year_counts = train['release_year'].value_counts()

years_sorted = sorted(year_counts.index)
sns.countplot(x='release_year', data=train, order=years_sorted)
plt.title("Movie releases by year")
plt.xticks(fontsize=12,rotation=90)
plt.show()

In [ ]:
plt.figure(figsize=(18,12))
year_counts = train['release_month'].value_counts()

years_sorted = sorted(year_counts.index)
sns.countplot(x='release_month', data=train, order=years_sorted)
plt.title("Release Month Count")
loc, labels = plt.xticks()
loc, labels = loc, ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
plt.xticks(loc, labels,fontsize=20)
plt.show()

In [ ]:
plt.figure(figsize=(20,12))
year_counts = train['release_day'].value_counts()

years_sorted = sorted(year_counts.index)
sns.countplot(x='release_day', data=train, order=years_sorted)
plt.title("Release Day Count")
plt.xticks(fontsize=20)
plt.show()

In [ ]:
plt.figure(figsize=(18,12))
year_counts = train['release_dayofweek'].value_counts()

years_sorted = sorted(year_counts.index)
sns.countplot(x='release_dayofweek', data=train, order=years_sorted)
plt.title("Total movies released on Day Of Week",fontsize=20)
loc, labels = plt.xticks()
loc, labels = loc, ["Mon","Tue","Wed","Thu","Fri","Sat","Sun"]
plt.xticks(loc, labels,fontsize=20)
plt.show()

In [ ]:
plt.figure(figsize=(20,12))
year_counts = train['release_quarter'].value_counts()

years_sorted = sorted(year_counts.index)
sns.countplot(x='release_quarter', data=train, order=years_sorted)
plt.title("Total movies released in a quarter",fontsize=20)
plt.xticks(ticks=[0,1,2,3],labels=['Jan-Mar','Apr-June','july-Sept','oct-dec'])
plt.show()

In [ ]:
plt.figure(figsize= (18,12))
sns.lineplot(x='release_year',y='revenue',data = train)
plt.title("Release year vs Revenue")
plt.show()

In [ ]:
plt.figure(figsize= (18,12))
sns.lineplot(x='release_year',y='budget',data = train,label = 'budget')
sns.lineplot(x='release_year',y='revenue',data = train,label = 'revenue')
plt.legend()
plt.show()

In [ ]:
plt.figure(figsize= (18,12))
sns.lineplot(x='release_month',y='revenue',data = train)
plt.title("Release month vs Revenue")
plt.show()

In [ ]:
plt.figure(figsize= (18,12))
sns.lineplot(x='release_dayofweek',y='revenue',data = train)
plt.title("Release day vs Revenue")
plt.show()

In [ ]:
plt.figure(figsize = (18,12))
sns.countplot(x='original_language', data=train)
plt.show()

In [ ]:
train['status'].value_counts()

In [ ]:
train['is_released'] = train['status'].apply(lambda x: 1 if x == 'Released' else 0)
test['is_released'] = test['status'].apply(lambda x: 1 if x == 'Released' else 0)

only 4 movies are rumored the rest are released

In [ ]:
train['has_homepage'] = train['homepage'].apply(lambda x: 0 if pd.isna(x) else 1)
test['has_homepage'] = test['homepage'].apply(lambda x: 0 if pd.isna(x) else 1)

In [ ]:
train = train.drop(['homepage'],axis = 1)
test = test.drop(['homepage'],axis = 1)

In [ ]:
plt.figure(figsize = (18,12))
sns.countplot(x='has_homepage', data=train,palette='winter')
plt.show()

most of the movies dont have a homepage

Does having a homepage effect the revenue

In [ ]:
train['has_a_tagline'] = train['tagline'].apply(lambda x: 0 if pd.isna(x) else 1)
test['has_a_tagline'] = test['tagline'].apply(lambda x: 0 if pd.isna(x) else 1)

In [ ]:
train['originallang_en'] = train['original_language'].apply(lambda x: 1 if x == 'en'else 0)
test['originallang_en'] = test['original_language'].apply(lambda x: 1 if x == 'en'else 0)

In [ ]:
plt.figure(figsize = (18,12))
sns.heatmap(train.corr(),annot = True,cmap = 'Blues');

## Lets start with 'belongs_to_collection' column

counting missing values

In [ ]:
train['belongs_to_collection'].apply(lambda x: len(x) if x != {} else 0).value_counts()

2396 values in this column are empty, 604 contrain information about the collections.

Adding a collection_name column which contains name of the collection (for eg: 'Hot Tub Time Machine Collection')

In [ ]:
train['collection_name'] = train['belongs_to_collection'].apply(lambda x: x[0]['name'] if x != {} else '')

Adding a collection_name column which will tell if the movie belongs to a specific colllection<br>
1 : if movie belongs to a collection<br>
0 : if it does'nt belong to any collection

In [ ]:
train['has_collection'] = train['belongs_to_collection'].apply(lambda x: len(x) if x != {} else 0)

Doing the same for the test set

In [ ]:
test['collection_name'] = test['belongs_to_collection'].apply(lambda x: x[0]['name'] if x != {} else '')
test['has_collection'] = test['belongs_to_collection'].apply(lambda x: len(x) if x != {} else 0)

Now we dont need the 'belongs_to_collection' column do we will drop it in both the sets

In [ ]:
train = train.drop(['belongs_to_collection'], axis=1)
test = test.drop(['belongs_to_collection'], axis=1)

## Now lets deal with then 'genres' column
one movie can have multiple genres

How many genres do films have

In [ ]:
train['genres'].apply(lambda x: len(x) if x != {} else 0).value_counts()

Now lets make a list of genres to find out which genres are the most popular

In [ ]:
genre_list = list(train['genres'].apply(lambda x: [i['name'] for i in x] if x != {} else []))

In [ ]:
genre_list

We can plot a word cloud to show the genres which are the most popular

In [ ]:
Counter([i for j in genre_list for i in j])

These are the most watched genres

Lets create a column which shows how many genres does a movie belongs to

In [ ]:
train['num_genres'] = train['genres'].apply(lambda x:len(x) if x!= {} else 0)

encoding the genres which appear the most

In [ ]:
train['all_genres'] = train['genres'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')

In [ ]:
top_genres =[ m[0] for m in Counter([i for j in genre_list for i in j]).most_common(20)]
for g in top_genres:
    train['genre_' + g] = train['all_genres'].apply(lambda x: 1 if g in x else 0)

Now doing the same for the test set

In [ ]:
test['num_genres'] = test['genres'].apply(lambda x: len(x) if x != {} else 0)
test['all_genres'] = test['genres'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
for g in top_genres:
    test['genre_' + g] = test['all_genres'].apply(lambda x: 1 if g in x else 0)


In [ ]:
train = train.drop(["genres"],axis=1)
test = test.drop(["genres"],axis=1)

We dont need the genre column any more so we can drop it from both the sets

## overview

How many production companies are involved in making films

In [ ]:
train['production_companies'].apply(lambda x: len(x) if x!= {} else 0 ).value_counts()

Looks like most of the movies have 1 or 2 production companies

In [ ]:
company_list = list(train['production_companies'].apply(lambda x:[i['name'] for i in x] if x != {} else []).values)
print(company_list)

In [ ]:
Counter([i for j in company_list for i in j]).most_common(30)

making a column for the number of companies involved in making the movie

In [ ]:
train['num_companies'] = train['production_companies'].apply(lambda x: len(x) if x != {} else 0)

Encoding the top 30 companies

In [ ]:
train['all_production_companies'] = train['production_companies'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
top_companies = [m[0] for m in Counter([i for j in company_list for i in j]).most_common(30)]
for g in top_companies:
    train['production_company_' + g] = train['all_production_companies'].apply(lambda x: 1 if g in x else 0)

Same for the test set

In [ ]:
test['num_companies'] = test['production_companies'].apply(lambda x: len(x) if x != {} else 0)
test['all_production_companies'] = test['production_companies'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
for g in top_companies:
    test['production_company_' + g] = test['all_production_companies'].apply(lambda x: 1 if g in x else 0)

now we can drop the 'production_companies' column

In [ ]:
train = train.drop(['production_companies', 'all_production_companies'], axis=1)
test = test.drop(['production_companies', 'all_production_companies'], axis=1)

## production_countries column

In [ ]:
print('Number of production countries in films')
train['production_countries'].apply(lambda x: len(x) if x != {} else 0).value_counts()

In [ ]:
list_of_countries = list(train['production_countries'].apply(lambda x: [i['name'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_countries for i in j]).most_common(25)

In [ ]:
train['num_countries'] = train['production_countries'].apply(lambda x: len(x) if x != {} else 0)


Encoding the top 25 countries

In [ ]:
train['all_countries'] = train['production_countries'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
top_countries = [m[0] for m in Counter([i for j in list_of_countries for i in j]).most_common(25)]
for g in top_countries:
    train['production_country_' + g] = train['all_countries'].apply(lambda x: 1 if g in x else 0)

same for the test set

In [ ]:
test['num_countries'] = test['production_countries'].apply(lambda x: len(x) if x != {} else 0)
test['all_countries'] = test['production_countries'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
for g in top_countries:
    test['production_country_' + g] = test['all_countries'].apply(lambda x: 1 if g in x else 0)

Dropping the 'production_countries' column

In [ ]:
train = train.drop(['production_countries', 'all_countries'], axis=1)
test = test.drop(['production_countries', 'all_countries'], axis=1)

## spoken_languages column

In [ ]:
print('Number of spoken languages in films')
train['spoken_languages'].apply(lambda x: len(x) if x != {} else 0).value_counts()

list of most common languages

In [ ]:
list_of_languages = list(train['spoken_languages'].apply(lambda x: [i['name'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_languages for i in j]).most_common()

Number of languages spoken in movies

In [ ]:
train['num_languages'] = train['spoken_languages'].apply(lambda x: len(x) if x != {} else 0)

Encoding the most common languages

In [ ]:
train['all_languages'] = train['spoken_languages'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
top_languages = [m[0] for m in Counter([i for j in list_of_languages for i in j]).most_common(30)]
for g in top_languages:
    train['language_' + g] = train['all_languages'].apply(lambda x: 1 if g in x else 0)

In [ ]:
test['num_languages'] = test['spoken_languages'].apply(lambda x: len(x) if x != {} else 0)
test['all_languages'] = test['spoken_languages'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
for g in top_languages:
    test['language_' + g] = test['all_languages'].apply(lambda x: 1 if g in x else 0)

Dropping the 'spoken_languages' column 

In [ ]:
train = train.drop(['spoken_languages', 'all_languages'], axis=1)
test = test.drop(['spoken_languages', 'all_languages'], axis=1)

## keywords

In [ ]:
print('Number of Keywords in films')
train['Keywords'].apply(lambda x: len(x) if x != {} else 0).value_counts().head(10)

In [ ]:
list_of_keywords = list(train['Keywords'].apply(lambda x: [i['name'] for i in x] if x != {} else []).values)

Column for number of keywords

In [ ]:
train['num_Keywords'] = train['Keywords'].apply(lambda x: len(x) if x != {} else 0)

encoding for most common keywords

In [ ]:
train['all_Keywords'] = train['Keywords'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
top_keywords = [m[0] for m in Counter([i for j in list_of_keywords for i in j]).most_common(30)]
for g in top_keywords:
    train['keyword_' + g] = train['all_Keywords'].apply(lambda x: 1 if g in x else 0)

In [ ]:
test['num_Keywords'] = test['Keywords'].apply(lambda x: len(x) if x != {} else 0)
test['all_Keywords'] = test['Keywords'].apply(lambda x: ' '.join(sorted([i['name'] for i in x])) if x != {} else '')
for g in top_keywords:
    test['keyword_' + g] = test['all_Keywords'].apply(lambda x: 1 if g in x else 0)

Dropping the Keywords column

In [ ]:
train = train.drop(['Keywords', 'all_Keywords'], axis=1)
test = test.drop(['Keywords', 'all_Keywords'], axis=1)

## Cast

In [ ]:
print('Number of casted persons in films')
train['cast'].apply(lambda x: len(x) if x != {} else 0).value_counts().head(10)

Most popular ones are

In [ ]:
list_of_cast_names = list(train['cast'].apply(lambda x: [i['name'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_cast_names for i in j]).most_common(15)

Gender of the cast

0 is unspecified, 1 is female, and 2 is male

In [ ]:
list_of_cast_genders = list(train['cast'].apply(lambda x: [i['gender'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_cast_genders for i in j]).most_common()

Characters played

In [ ]:
list_of_cast_characters = list(train['cast'].apply(lambda x: [i['character'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_cast_characters for i in j]).most_common(15)

Creating number column for the cast

In [ ]:
train['num_cast'] = train['cast'].apply(lambda x: len(x) if x != {} else 0)

In [ ]:
plt.figure(figsize=(16, 8))
plt.scatter(train['num_cast'], train['revenue'])
plt.title('Number of cast members vs revenue');

Encoding the top cast names

In [ ]:
top_cast_names = [m[0] for m in Counter([i for j in list_of_cast_names for i in j]).most_common(15)]
for g in top_cast_names:
    train['cast_name_' + g] = train['cast'].apply(lambda x: 1 if g in str(x) else 0)

Encoding the cast according to their Genders

In [ ]:
train['genders_0_cast'] = train['cast'].apply(lambda x: sum([1 for i in x if i['gender'] == 0]))
train['genders_1_cast'] = train['cast'].apply(lambda x: sum([1 for i in x if i['gender'] == 1]))
train['genders_2_cast'] = train['cast'].apply(lambda x: sum([1 for i in x if i['gender'] == 2]))

Encoding the characters played

In [ ]:
top_cast_characters = [m[0] for m in Counter([i for j in list_of_cast_characters for i in j]).most_common(15)]
for g in top_cast_characters:
    train['cast_character_' + g] = train['cast'].apply(lambda x: 1 if g in str(x) else 0)

Doing the same for the test set

In [ ]:
test['num_cast'] = test['cast'].apply(lambda x: len(x) if x != {} else 0)
for g in top_cast_names:
    test['cast_name_' + g] = test['cast'].apply(lambda x: 1 if g in str(x) else 0)
test['genders_0_cast'] = test['cast'].apply(lambda x: sum([1 for i in x if i['gender'] == 0]))
test['genders_1_cast'] = test['cast'].apply(lambda x: sum([1 for i in x if i['gender'] == 1]))
test['genders_2_cast'] = test['cast'].apply(lambda x: sum([1 for i in x if i['gender'] == 2]))

In [ ]:
for g in top_cast_characters:
    test['cast_character_' + g] = test['cast'].apply(lambda x: 1 if g in str(x) else 0)

Dropping the cast column

In [ ]:
train = train.drop(['cast'], axis=1)
test = test.drop(['cast'], axis=1)

## Crew column

In [ ]:
print('Number of crew members')
train['crew'].apply(lambda x: len(x) if x != {} else 0).value_counts().head(10)

Number of crew members


2     179
11    127
10    126
3     126
12    110
8     109
9     109
14    104
4     101
7      94
Name: crew, dtype: int64

In [ ]:
list_of_crew_names = list(train['crew'].apply(lambda x: [i['name'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_crew_names for i in j]).most_common(15)

[('Avy Kaufman', 50),
 ('Robert Rodriguez', 44),
 ('Deborah Aquila', 40),
 ('James Newton Howard', 39),
 ('Mary Vernieu', 38),
 ('Steven Spielberg', 37),
 ('Luc Besson', 37),
 ('Jerry Goldsmith', 37),
 ('Francine Maisler', 35),
 ('Tricia Wood', 35),
 ('James Horner', 33),
 ('Kerry Barden', 32),
 ('Bob Weinstein', 30),
 ('Harvey Weinstein', 30),
 ('Janet Hirshenson', 30)]

Crew jobs

In [ ]:
list_of_crew_jobs = list(train['crew'].apply(lambda x: [i['job'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_crew_jobs for i in j]).most_common(15)

[('Producer', 6011),
 ('Executive Producer', 3459),
 ('Director', 3225),
 ('Screenplay', 2996),
 ('Editor', 2824),
 ('Casting', 2483),
 ('Director of Photography', 2288),
 ('Original Music Composer', 1947),
 ('Art Direction', 1821),
 ('Production Design', 1650),
 ('Costume Design', 1573),
 ('Writer', 1523),
 ('Set Decoration', 1345),
 ('Makeup Artist', 1108),
 ('Sound Re-Recording Mixer', 970)]

Crew genders

In [ ]:
list_of_crew_genders = list(train['crew'].apply(lambda x: [i['gender'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_crew_genders for i in j]).most_common(15)

[(0, 41787), (2, 24898), (1, 6412)]

In [ ]:
list_of_crew_departments = list(train['crew'].apply(lambda x: [i['department'] for i in x] if x != {} else []).values)
Counter([i for j in list_of_crew_departments for i in j]).most_common(14)

[('Production', 15887),
 ('Sound', 9319),
 ('Art', 8069),
 ('Crew', 7315),
 ('Writing', 6567),
 ('Costume & Make-Up', 6156),
 ('Camera', 5424),
 ('Directing', 4954),
 ('Editing', 4508),
 ('Visual Effects', 3591),
 ('Lighting', 1303),
 ('Actors', 4)]

Creating Number of crew members column

In [ ]:
train['num_crew'] = train['crew'].apply(lambda x: len(x) if x != {} else 0)

In [ ]:
plt.figure(figsize=(16, 8))
plt.scatter(train['num_crew'], train['revenue'])
plt.title('Number of crew members vs revenue');

In [ ]:
top_crew_names = [m[0] for m in Counter([i for j in list_of_crew_names for i in j]).most_common(15)]
for g in top_crew_names:
    train['crew_name_' + g] = train['crew'].apply(lambda x: 1 if g in str(x) else 0)

In [ ]:
train['genders_0_crew'] = train['crew'].apply(lambda x: sum([1 for i in x if i['gender'] == 0]))
train['genders_1_crew'] = train['crew'].apply(lambda x: sum([1 for i in x if i['gender'] == 1]))
train['genders_2_crew'] = train['crew'].apply(lambda x: sum([1 for i in x if i['gender'] == 2]))

Encoding the most common crew jobs and departments

In [ ]:
top_crew_jobs = [m[0] for m in Counter([i for j in list_of_crew_jobs for i in j]).most_common(15)]
for j in top_crew_jobs:
    train['jobs_' + j] = train['crew'].apply(lambda x: sum([1 for i in x if i['job'] == j]))
top_crew_departments = [m[0] for m in Counter([i for j in list_of_crew_departments for i in j]).most_common(15)]
for j in top_crew_departments:
    train['departments_' + j] = train['crew'].apply(lambda x: sum([1 for i in x if i['department'] == j])) 

Applying the same to the test set

In [ ]:
test['num_crew'] = test['crew'].apply(lambda x: len(x) if x != {} else 0)
for g in top_crew_names:
    test['crew_name_' + g] = test['crew'].apply(lambda x: 1 if g in str(x) else 0)
test['genders_0_crew'] = test['crew'].apply(lambda x: sum([1 for i in x if i['gender'] == 0]))
test['genders_1_crew'] = test['crew'].apply(lambda x: sum([1 for i in x if i['gender'] == 1]))
test['genders_2_crew'] = test['crew'].apply(lambda x: sum([1 for i in x if i['gender'] == 2]))
for j in top_crew_jobs:
    test['jobs_' + j] = test['crew'].apply(lambda x: sum([1 for i in x if i['job'] == j]))
for j in top_crew_departments:
    test['departments_' + j] = test['crew'].apply(lambda x: sum([1 for i in x if i['department'] == j])) 

In [ ]:
# from sklearn.preprocessing import OneHotEncoder

In [ ]:
# cat_cols = ['original_language','collection_name']

In [ ]:
# encoder = OneHotEncoder(sparse=False, handle_unknown='ignore')
# encoder.fit(train[cat_cols])

In [ ]:
# encoded_cols = list(encoder.get_feature_names_out(cat_cols))
# print(encoded_cols)

In [ ]:
# new_train = train.copy()
# new_test = test.copy()

In [ ]:
# new_train[encoded_cols] = encoder.transform(new_train[cat_cols])
# new_test[encoded_cols] = encoder.transform(new_test[cat_cols])

In [ ]:
# train = new_train.copy()
# test = new_test.copy()

Adding additional Features

In [ ]:
train['bud_runtime']=(train['budget']+1)/(train['runtime']+1)

train['bud_year']=(train['budget']+1)/(train['release_year']+1)

train['bud_popularity']=(train['budget']+1)/(train['popularity']+1)

train['runtime_year']=(train['runtime']+1)/(train['release_year']+1)

train['popularity_year']=(train['popularity']+1)/(train['release_year']+1)

In [ ]:
test['bud_runtime']=(test['budget']+1)/(test['runtime']+1)

test['bud_year']=(test['budget']+1)/(test['release_year']+1)

test['bud_popularity']=(test['budget']+1)/(test['popularity']+1)

test['runtime_year']=(test['runtime']+1)/(test['release_year']+1)

test['popularity_year']=(test['popularity']+1)/(test['release_year']+1)

Dropping a few columns

In [ ]:
train = train.drop(['crew'], axis=1)
test = test.drop(['crew'], axis=1)

In [ ]:
train = train.drop(['imdb_id','release_date', 'status'], axis=1)
test = test.drop(['imdb_id','release_date', 'status'], axis=1)

In [ ]:
train = train.drop(['all_genres'], axis=1)
test = test.drop(['all_genres'], axis=1)

In [ ]:
train.shape, test.shape

In [ ]:
X = train

In [ ]:
corr_matrix = X.corr()
sns.heatmap(corr_matrix,cmap='coolwarm')
plt.show()
imp_features = []
for feature in corr_matrix.columns:
  if abs(corr_matrix[feature]['revenue']) >= 0.3:
    imp_features.append(feature)


# Show the plot
plt.show()
# print(imp_features)
X = X[imp_features]


In [ ]:
imp_features.remove('revenue')
test = test[imp_features]

In [ ]:
X.shape

In [ ]:
trained_model = X
for i in ['original_language','overview','tagline','title','id','original_title']:
  if i in trained_model.columns:
    trained_model = trained_model.drop(i,axis = 1)

In [ ]:
trained_model
trained_model.to_csv('trainpreprocess.csv', index=False)

In [ ]:
X.head()

## Creating Models

In [ ]:
y = train['revenue']

In [ ]:
for i in ['original_language','overview','tagline','title','id','revenue','original_title']:
  if i in X.columns:
    X = X.drop(i,axis = 1)


In [ ]:
for i in ['original_language','overview','tagline','title','id','original_title']:
  if i in test.columns:
    test = test.drop(i,axis = 1)

In [ ]:
X.shape,y.shape,test.shape

In [ ]:
X.columns

In [ ]:
train

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
X_train,X_val,y_train,y_val= train_test_split(X,y,test_size=0.3,random_state=42)

In [ ]:
X_train.shape,y_train.shape

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn import metrics
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3, random_state=42)

# Define the hyperparameters to be tuned
param_grid = {
    'fit_intercept': [True, False],
}

# Define the linear regression model
from sklearn.metrics import confusion_matrix,f1_score,accuracy_score
lr = LinearRegression()

# Define the grid search object
grid_search = GridSearchCV(lr, param_grid=param_grid, cv=5)

# Fit the grid search object to the training data
grid_search.fit(X_train, y_train)

# Print the best hyperparameters and corresponding R-squared score
print("Best Hyperparameters: ", grid_search.best_params_)
print("Best R-squared Score: ", grid_search.best_score_)

# Get the best model and make predictions on the validation set
best_model = grid_search.best_estimator_
lr_pred = best_model.predict(X_val)
print('MAE:', metrics.mean_absolute_error(lr_pred, y_val))
print('MSE:', metrics.mean_squared_error(lr_pred, y_val))
linear_rmse = np.sqrt(metrics.mean_squared_error(lr_pred, y_val))
grid_search.fit(X,y)
best_model = grid_search.best_estimator_
lr_full = best_model.predict(test)


In [ ]:
from sklearn.metrics import r2_score
r2score = r2_score(y_val, lr_pred)
print(r2score)

In [ ]:
#REGULARIZATION METHOD
from sklearn.linear_model import Lasso
lasso_model = Lasso(alpha=0.001)

# Fit the Lasso model to the training data
lasso_model.fit(X_train, y_train)

# Make predictions on the testing data
ls_pred = lasso_model.predict(X_val)

# Calculate the R-squared score on the testing data
r2 = r2_score(y_val, ls_pred)

print("Lasso Regression R-squared score:", r2)
print('MAE:', metrics.mean_absolute_error(ls_pred, y_val))
print('MSE:', metrics.mean_squared_error(ls_pred, y_val))
lasso_rmse = np.sqrt(metrics.mean_squared_error(ls_pred, y_val))

In [ ]:
from sklearn.linear_model import Ridge

# Define the parameter grid to search over
param_grid = {
    'fit_intercept': [True, False],
    'alpha': [0.1, 0.5, 1, 5, 10, 50, 100]
}


# Define the Ridge regression model
rr = Ridge()

# Define the grid search object
grid_search = GridSearchCV(rr, param_grid=param_grid, cv=5)

# Fit the grid search object to the training data
grid_search.fit(X_train, y_train)
best_model = grid_search.best_estimator_
rd_pred = best_model.predict(X_val)

# Print the best hyperparameters and corresponding R-squared score
print("Best Hyperparameters: ", grid_search.best_params_)
print("Best R-squared Score: ", grid_search.best_score_)
print('MAE:', metrics.mean_absolute_error(rd_pred, y_val))
print('MSE:', metrics.mean_squared_error(rd_pred, y_val))
rideg_rmse = np.sqrt(metrics.mean_squared_error(rd_pred, y_val))


In [ ]:
ridge = Ridge(alpha=2000)

# Fit the model to the training data
ridge.fit(X_train, y_train)

# Predict the target values for the test data
y_pred = ridge.predict(X_val)
print('MAE:', metrics.mean_absolute_error(y_pred, y_val))
print('MSE:', metrics.mean_squared_error(y_pred, y_val))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_pred, y_val)))

In [ ]:
# lr_pred=linreg.predict(X_val)

In [ ]:
lr_rmse= mean_squared_error(y_val, lr_pred, squared=False)
lr_rmse

In [ ]:
X.isna().sum()

In [ ]:
lr_full

In [ ]:
lr_full = np.array(lr_full, dtype=np.float128)

In [ ]:
lr_full = np.expm1(lr_full)

In [ ]:
submission['revenue'] = lr_full
submission.to_csv("lr.csv", index=False)

In [ ]:
submission.head()

In [ ]:
from sklearn import metrics

In [ ]:
print('MAE:', metrics.mean_absolute_error(lr_pred, y_val))
print('MSE:', metrics.mean_squared_error(lr_pred, y_val))
print('RMSE:', np.sqrt(metrics.mean_squared_error(lr_pred, y_val)))

In [ ]:
X_train.shape

## RandomForest

In [ ]:
from sklearn.ensemble import RandomForestRegressor
rf_model = RandomForestRegressor(random_state=42, 
                                  max_features='auto', 
                                  n_estimators= 1000, 
                                  min_samples_leaf=5)

In [ ]:
rf_model.fit(X_train,y_train)

In [ ]:
rf_pred = rf_model.predict(X_val)

In [ ]:
rf_full = rf_model.predict(test)

In [ ]:
rf_full = np.array(rf_full, dtype=np.float128)

In [ ]:
rf_rmse = mean_squared_error(y_val, rf_pred, squared=False)
rf_rmse

In [ ]:
rf_full = np.expm1(rf_full)

In [ ]:
submission['revenue'] =rf_full 
submission.to_csv("rf.csv", index=False)

In [ ]:
print('MAE:', metrics.mean_absolute_error(rf_pred, y_val))
print('MSE:', metrics.mean_squared_error(rf_pred, y_val))
random_forest_rmse = np.sqrt(metrics.mean_squared_error(rf_pred, y_val))

## XGB

In [ ]:
!pip install xgboost --upgrade --quiet

In [ ]:
# def XGBoost(X,y,file):
#   from sklearn.model_selection import train_test_split
#   X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3, random_state=42)

from xgboost import XGBRegressor
xgb = XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=3)

  # Fit the model to the training data
xgb.fit(X_train, y_train)

  # Make predictions on the validation set
xgb_pred = xgb.predict(X_val)
xgb.fit(X,y)
xgb_full = xgb.predict(test)
xgb_full = np.array(xgb_full, dtype=np.float128)
xgb_full = np.expm1(xgb_full)
submission['revenue'] = xgb_full
submission.to_csv("xgb.csv", index=False)
# return submission
mse = mean_squared_error(y_val, xgb_pred)
print("Mean Squared Error: ", mse)

In [ ]:

xgb = XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=3)

# Fit the model to the training data
xgb.fit(X_train, y_train)

# Make predictions on the validation set
xgb_pred = xgb.predict(X_val)

# Evaluate the model using mean squared error
mse = mean_squared_error(y_val, xgb_pred)
print("Mean Squared Error: ", mse)

In [ ]:
xgb = XGBRegressor()

# Define the grid search parameters
param_grid = {
    'learning_rate': [0.01, 0.05, 0.1, 0.2],
    'max_depth': [3, 5, 7],
    'n_estimators': [50, 100, 200],
    'subsample': [0.5, 0.7, 1],
}

# Define the grid search object
grid_search = GridSearchCV(xgb, param_grid=param_grid, cv=5, n_jobs=-1, verbose=1)

# Fit the grid search object to the training data
grid_search.fit(X_train, y_train)

# Print the best hyperparameters and corresponding R-squared score
print("Best Hyperparameters: ", grid_search.best_params_)
print("Best R-squared Score: ", grid_search.best_score_)

# Make predictions on the validation set using the best model
best_model = grid_search.best_estimator_
xgb_pred = best_model.predict(X_val)

# Evaluate the model using mean squared error
mse = mean_squared_error(y_val, xgb_pred)
print("Mean Squared Error: ", mse)

In [ ]:
print('MAE:', metrics.mean_absolute_error(xgb_pred, y_val))
print('MSE:', metrics.mean_squared_error(xgb_pred, y_val))
xgb_rmse = np.sqrt(metrics.mean_squared_error(xgb_pred, y_val))

In [ ]:
r2score = r2_score(y_val, lr_pred)
print(r2score)

In [ ]:
data = {'Model': ['Linear Regression', 'Lasso Regression', 'Ridge Regression', 'Random Forest', 'XGBoost'], 'RMSE': [linear_rmse, lasso_rmse, rideg_rmse, random_forest_rmse, xgb_rmse]}
df = pd.DataFrame(data)
df.head()

In [ ]:
rmse_values = [linear_rmse, lasso_rmse, rideg_rmse, random_forest_rmse, xgb_rmse]

# Algorithm names
algorithm_names = ['Linear', 'Lasso', 'Ridge', 'Random Forest', 'XGBoost']

# Plot the RMSE values as a line graph
plt.plot(algorithm_names, rmse_values)

# Label the axes and add a title
# plt.xlabel('Algorithm')
plt.ylabel('RMSE')
plt.title('RMSE Values for Different Algorithms')

# Display the graph
plt.show()

In [ ]:
!pip install gradio
import gradio as gr
def movie_critic(budget, has_a_tagline,production_country_United_States_of_America, num_cast,genders_2_cast, num_crew, genders_1_crew, genders_2_crew,departments_Sound, departments_Art, departments_Editing,bud_year):
    data = pd.DataFrame({
        'budget': [budget],
        'has_a_tagline': [has_a_tagline],
        'production_country_United States of America': [production_country_United_States_of_America],
        'num_cast': [num_cast],
        'genders_2_cast': [genders_2_cast],
        'num_crew': [num_crew],
        'genders_1_crew': [genders_1_crew],
        'genders_2_crew': [genders_2_crew],
        'departments_Sound': [departments_Sound],
        'departments_Art': [departments_Art],
        'departments_Editing': [departments_Editing],
        'bud_year': [bud_year]
    })
    from sklearn.model_selection import train_test_split
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3, random_state=42)
    from xgboost import XGBRegressor
    xgb = XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=3)

    # Fit the model to the training data
    xgb.fit(X_train, y_train)

    # Make predictions on the validation set
    xgb.fit(X,y)
    prediction = xgb.predict(data)
    # prediction = np.array(xgb_full, dtype=np.float128)
    prediction = np.expm1(prediction)
    return prediction[0]
    

# Create the Gradio interface
gradio_interface = gr.Interface(
    fn=movie_critic,
    inputs=[
        gr.inputs.Number(label='Budget in 10 million'),
        gr.inputs.Number(label='Does the movie has tagline?'),
        gr.inputs.Number(label='is the production country United States of America?'),
        gr.inputs.Number(label='Number of cast members'),
        gr.inputs.Number(label='Number of male cast members'),
        gr.inputs.Number(label='Number of crew members'),
        gr.inputs.Number(label='No of female crew members'),
        gr.inputs.Number(label='No of male crew members'),
        gr.inputs.Number(label='No.of members in sound department'),
        gr.inputs.Number(label='No.of members in art department'),
        gr.inputs.Number(label='No.of members in editing department'),
        gr.inputs.Number(label='Enter the ratio of budget and release year'),

    ],
      outputs=gr.outputs.Textbox(label='revenue'),
      title = "Revenue prediction",
      layout="vertical",
      theme = "compact"
)

# Launch the interface
gradio_interface.launch(debug=False,share=True)